Fix navigation accessibility issues

toolness opened the pull request:
This fixes a number of accessibility-related bugs with navigation:

* Fixes #2029 
* Fixes #1653 
* Fixes #1649

It also addresses a few other focus issues that are documented in the GitHub/code comments below.

To do:
- [x] Add tests.
- [x] ~~Consider ensuring that keyboard focus can't leave the navigation menu once it's activated. The menu essentially seems to function like a dialog box, and behaves much like one from a mouse/touch standpoint, so we might want to make sure it behaves like one from a keyboard standpoint as well.~~ Filed separately as #2031.

toolness commented on src/stylesheets/components/_navigation.scss:
```diff
@@ -62,11 +63,14 @@
   &.is-visible {
     @include transform(translateX(0));
-    @include transition(all 0.3s ease-in-out);
+    @include transition(transform 0.3s ease-in-out);
```
Changing the transition from `all` to just `transform` was crucial for ensuring that we could still focus on the close button immediately after the user clicked on the menu button: with transition set to `all`, even `visibility` was being transitioned, which meant that when `focus()` was called on the close button, nothing happened because the button wasn't actually visible yet!

toolness commented on src/stylesheets/components/_navigation.scss:
```diff
@@ -51,6 +51,7 @@
   padding: 2rem;
   width: $sliding-panel-width;
   z-index: $z-index-nav;
+  visibility: hidden;
```
This ensures that keyboard focus can't go into the nav menu unless it's visible, either because desktop mode is active, or because the mobile nav is active.

toolness commented on src/js/components/navigation.js:
```diff
@@ -7,7 +7,6 @@ const accordion = require('./accordion');
 const CLICK = require('../events').CLICK;
 const PREFIX = require('../config').prefix;
 
-const CONTEXT = 'header';
```
I got rid of this and the focus logic that required the close button be inside it because the close button *isn't* inside a `<header>` on standards.usa.gov, which meant that the close button wasn't even being focused on that site after the user activated the menu. 😞

toolness commented:
Ok, I think this is ready for review. I've manually tested it on Chrome, Firefox, and IE11 and it seems to work fine on them.

msecret commented on spec/unit/navigation/navigation.spec.js:
```diff
@@ -66,6 +67,39 @@ describe('navigation toggle', function () {
+    try {
+      window.dispatchEvent(new CustomEvent('resize'));
+    } finally {
+      closeButton.getBoundingClientRect.restore();
```
Rather then doing this in a finally, you can use the sinon sandbox and restore in in an afterEach. Then you wouldn't need both of these. Or is there something else about the `try` that's beneficial?

toolness replied:
Ah nice, I didn't know about [sinon sandboxes](http://sinonjs.org/releases/v1.17.7/sandbox/)!  Thanks, I'll convert this to use them.

msecret commented on src/js/components/navigation.js:
```diff
@@ -26,20 +25,43 @@ const toggleNav = function (active) {
 };
 
+const resize = () => {
+  const isActive = document.body.classList.contains(ACTIVE_CLASS);
```
Shouldn't we have a util for this?

msecret replied:
Or I guess we're not, seems pretty simple so maybe it's fine

toolness commented:
Ok, I just resolved your concerns and also fixed a merge conflict due to #2006.  I also added some comments around some of the CSS changes I've made, because I think they're particularly non-obvious, in that they don't change the behavior of the site for *sighted* users, but do affect its usability for keyboard and AT users.
